crates/pallet-evm-system/src/migrations/broken_nonces_recovery.rs

```diff
@@ -0,0 +1,189 @@
+//! Migration to recover broken nonces.
+
+#[cfg(feature = "try-runtime")]
+use frame_support::sp_std::vec::Vec;
+use frame_support::{log::info, pallet_prelude::*, traits::OnRuntimeUpgrade};
+use rlp::RlpStream;
+use sp_core::H160;
+use sp_io::hashing::keccak_256;
+use sp_runtime::traits::{CheckedAdd, One, Zero};
+
+use crate::{Account, AccountInfo, Config, Pallet};
+
+/// EVM state provider.
+pub trait EvmStateProvider<AccountId> {
+    /// Check whether account address has code.
+    fn has_code(account_id: &AccountId) -> (bool, Weight);
+
+    /// Gives an iterator over accounts managed by EVM. Considers precompiled contracts' accounts
+    /// as not "managed by EVM".
+    fn accounts_managed_by_evm() -> impl Iterator<Item = (AccountId, Weight)>;
+}
+
+/// Execute migration to recover broken nonces.
+pub struct MigrationBrokenNoncesRecovery<EP, T>(sp_std::marker::PhantomData<(EP, T)>);
+
+#[cfg(feature = "try-runtime")]
+#[derive(Encode, Decode)]
+/// Key indicators of the state before runtime upgrade.
+struct PreUpgradeState {
+    /// Accounts' count.
+    accounts: u64,
```

Comment:
Should we add check that account `data` value isn't changed during migration for each fixed account  managed by evm?

Reply:
I can assume that it's redundant, but it's better to ask :)